group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -61,6 +67,29 @@
  *    handling as well as during the initial loading of the records from the partitions.
  */
 public class OffsetMetadataManager {
+
+    /**
+     * An offset is considered expired based on different factors, such as the state of the group
+     * and/or the GroupMetadata record version (for generic groups). This class is used to check
+     * how offsets for the group should be expired.
+     */
+    public static class ExpirationCondition {
+        /**
+         * Given an offset metadata, return the base timestamp that will be used to calculate expiration.
+         */
+        final Function<OffsetAndMetadata, Long> baseTimestamp;
+
+        /**
+         * The set of subscribed topics a group should check against for an offset commit to be eligible for expiration.
+         */
+        final Set<String> subscribedTopics;
+
+        public ExpirationCondition(Function<OffsetAndMetadata, Long> baseTimestamp, Set<String> subscribedTopics) {
+            this.baseTimestamp = baseTimestamp;
+            this.subscribedTopics = subscribedTopics;
+        }
+    }
```

Comment:
As we don't need `subscribedTopics` here anymore, I think that we could simplify this. How about the following?

```
interface OffsetExpirationCondition {
  boolean isOffsetExpired(OffsetAndMetadata offset, long offsetsRetentionMs);
}
```

With this, you can define it with a lambda. This would be pretty convenient, I suppose.

```
(offset, offsetsRetentionMs) -> {
   ....
   return true/false;
}
```